yapapi/contrib/strategy/rep_a1.py

```diff
@@ -0,0 +1,107 @@
+from typing import Dict, Set, TYPE_CHECKING
+from decimal import Decimal
+from collections import defaultdict
+import logging
+from random import gauss
+
+from yapapi.strategy import WrappingMarketStrategy
+from yapapi import events
+
+if TYPE_CHECKING:
+    from yapapi.rest.payment import DebitNote, Invoice
+    from yapapi.rest.market import OfferProposal
+
+
+logger = logging.getLogger(__name__)
+
+ActivityId = str
+AgreementId = str
+
+
+async def get_provider_score(provider_id: str) -> float:
+    #   This is a mock.
+    #   In the final (alpha1) version this will query the API for the provider score,
+    #   but the score received will have the same "meaning", i.e.: how this provider
+    #   differs from the average provider quality, expressed in the number of SDs.
+    return gauss(0, 1)
+
+
+class RepA1(WrappingMarketStrategy):
+    def __init__(self, base_strategy):
+        super().__init__(base_strategy)
+
+        self._failed_activities: Set[ActivityId] = set()
+        self._accepted_amounts: Dict[ActivityId, Decimal] = defaultdict(Decimal)
+        self._agreement_activity_map: Dict[AgreementId, Set[ActivityId]] = defaultdict(set)
+
+    #################
+    #   OFFER SCORING
+    async def score_offer(self, offer: "OfferProposal") -> float:
+        offer_score = await super().score_offer(offer)
+        provider_score = await get_provider_score(offer.issuer)
+        combined_score = self._final_score(offer_score, provider_score)
+
+        provider_name = offer._proposal.proposal.properties['golem.node.id.name']
+        logger.info(
+            "Scored %s -  base: %s, provider: %s, combined: %s",
+            provider_name, offer_score, provider_score, combined_score
+        )
+        return combined_score
+
+    def _final_score(self, base_score: float, provider_score: float) -> float:
+        #   NOTE: this logic is just a POC
+        if provider_score < -1.5:
+            return -1
+        return base_score + provider_score
+
+    ######################
+    #   PAYMENT MANAGEMENT
+    def on_event(self, event: events.Event):
+        if isinstance(event, events.ActivityEvent):
+            self._agreement_activity_map[event.agreement.id].add(event.activity.id)
+
+        if isinstance(event, events.WorkerFinished):
+            if event.exception is not None:
+                self._activity_failed(event, "WORKER EXCEPTION")
```

Comment:
[potentially important] hmm... I'd test this with the `scan.py` example and see if we don't catch the `StopAsyncIteration` here by accident...